lib/validators/serialtiles.js

```diff
@@ -21,20 +22,73 @@ module.exports = function validateSerialtiles(opts, callback) {
     .on('tile', function(t) {
       if (errored) return;
       if (!t.buffer) return;
-      if (t.buffer.length <= limits.max_tilesize) return;
-      callback(invalid('Tile exceeds maximum size of ' + Math.round(limits.max_tilesize/1024) + 'k at z' + t.z + '. Reduce the detail of data at this zoom level or omit it by adjusting your minzoom.'));
-      errored = true;
-      read.close();
+
+      var gzipped = (tiletype.headers(t.buffer))['Content-Encoding'] === 'gzip';
+      var format = tiletype.type(t.buffer);
+
+      // Not gzipped and not a valid tiletype
+      if (!gzipped && (format !== 'png' && format !== 'webp' && format !== 'jpg')) {
+        callback(invalid('Invalid tiletype'));
+        errored = true;
+        read.close();
+      }
+      
+      // Gzipped pbf
+      if (gzipped) {
+        //gunzip buffer and validate
+        validateGzipped(t, function(err) {
+          if (err) {
+            callback(invalid('Invalid pbf: %s', err));
+            errored = true;
+            read.close();
+          }
+
+          if (t.buffer.length <= limits.max_tilesize) return;
+          else callback(invalid('Tile exceeds maximum size of ' + Math.round(limits.max_tilesize/1024) + 'k at z' + t.z + '. Reduce the detail of data at this zoom level or omit it by adjusting your minzoom.'));
+          errored = true;
+          read.close();
+        });
+        //else validate maxsize of png/webp/jpg
+      } else {
+        if (t.buffer.length <= limits.max_tilesize) return;
+        else callback(invalid('Tile exceeds maximum size of ' + Math.round(limits.max_tilesize/1024) + 'k at z' + t.z + '. Reduce the detail of data at this zoom level or omit it by adjusting your minzoom.'));
+        errored = true;
+        read.close();
+      }
     })
     .once('error', function(err) {
       if (!errored) {
-        if (err.name === 'DeserializationError') callback(invalid('%s: %s', err.name, err.message));
-        else callback(err);
+        if (err.name === 'DeserializationError') return callback(invalid('%s: %s', err.name, err.message));
+        else return callback(err);
       }
       errored = true;
     })
     .on('end', function() {
-      if (!errored) callback();
+      if (!errored) return callback();
     })
     .resume();
-};
+
+  function validateGzipped(tile, callback) {
+    var vtile;
+
+    try {
+      vtile = new mapnik.VectorTile(tile.z,tile.x,tile.y);
+    } catch (err) { throw err; }
+
+    zlib.gunzip(tile.buffer, function(err, data) { 
+      if (err) throw err;
+
+      vtile.setData(data, function(err) {
+        if (err) throw err;
+        
+        vtile.parse();
+        if (vtile.empty()) return callback(invalid('Tile is empty'));
+        if (!vtile.toJSON()[0].name) return callback(invalid('Tile has no layers'));
```

Comment:
Slightly concerned that if you were to hand in something invalid-ish, you might land some `undefined has no property 'name'` kinds of errors here. It might be smarter to

```js
var json = vtile.toJSON();
if (!json[0] || !json[0].name) return callback(invalid('Tile has no layers'));
if (!json[0] || !json[0].features) return callback(invalid('Tile has no features'));
```